sklearn/metrics/_classification.py

```diff
@@ -337,62 +337,76 @@ def confusion_matrix(
 
     if sample_weight is None:
-        sample_weight = np.ones(y_true.shape[0], dtype=np.int64)
+        sample_weight = xp.ones(y_true.shape[0], dtype=xp.int64, device=device_)
     else:
-        sample_weight = np.asarray(sample_weight)
+        sample_weight = xp.asarray(sample_weight, device=device_)
 
     check_consistent_length(y_true, y_pred, sample_weight)
 
-    n_labels = labels.size
+    # TODO: remove condition when torch supports the size attribute
+    if xp.__name__ == "array_api_compat.torch":
+        n_labels = xp.size(labels)
+    else:
+        n_labels = labels.size
     # If labels are not consecutive integers starting from zero, then
     # y_true and y_pred must be converted into index form
     need_index_conversion = not (
-        labels.dtype.kind in {"i", "u", "b"}
-        and np.all(labels == np.arange(n_labels))
-        and y_true.min() >= 0
-        and y_pred.min() >= 0
+        xp.isdtype(labels.dtype, ("signed integer", "unsigned integer", "bool"))
+        and xp.all(labels == xp.arange(n_labels, device=device_))
+        and xp.min(y_true) >= 0
+        and xp.min(y_pred) >= 0
     )
     if need_index_conversion:
         label_to_ind = {y: x for x, y in enumerate(labels)}
-        y_pred = np.array([label_to_ind.get(x, n_labels + 1) for x in y_pred])
-        y_true = np.array([label_to_ind.get(x, n_labels + 1) for x in y_true])
+        y_pred = xp.asarray(
+            [label_to_ind.get(x, n_labels + 1) for x in y_pred], device=device_
+        )
+        y_true = xp.asarray(
+            [label_to_ind.get(x, n_labels + 1) for x in y_true], device=device_
+        )
 
     # intersect y_pred, y_true with labels, eliminate items not in labels
-    ind = np.logical_and(y_pred < n_labels, y_true < n_labels)
-    if not np.all(ind):
+    ind = xp.logical_and(y_pred < n_labels, y_true < n_labels)
+    if not xp.all(ind):
         y_pred = y_pred[ind]
         y_true = y_true[ind]
         # also eliminate weights of eliminated items
         sample_weight = sample_weight[ind]
 
     # Choose the accumulator dtype to always have high precision
-    if sample_weight.dtype.kind in {"i", "u", "b"}:
-        dtype = np.int64
+    if xp.isdtype(sample_weight.dtype, ("signed integer", "unsigned integer", "bool")):
+        dtype = xp.int64
     else:
-        dtype = np.float64
-
-    cm = coo_matrix(
-        (sample_weight, (y_true, y_pred)),
-        shape=(n_labels, n_labels),
-        dtype=dtype,
-    ).toarray()
+        dtype = xp.float64
+
+    if _is_numpy_namespace(xp):
+        cm = coo_matrix(
+            (sample_weight, (y_true, y_pred)),
+            shape=(n_labels, n_labels),
+            dtype=dtype,
+        ).toarray()
+    else:
+        cm = xp.zeros((n_labels, n_labels), dtype=dtype, device=device_)
+        for true, pred, weight in zip(y_true, y_pred, sample_weight):
+            cm[true, pred] += weight
 
     with np.errstate(all="ignore"):
```

Comment:
Here, I did a quick research if any of the other array libraries (other than numpy) would raise these warnings for divisions by zero as well. Result: it doesn't seem they do.

But just to be sure, is there any need to handle these warning for any other array library?

Reply:
I don't think there is any standardization of warnings and exceptions in the array API standard at this point unfortunately:

https://data-apis.org/array-api/latest/design_topics/exceptions.html

Reply:
If it doesn't affect the code flow then I think it should be fine. However @ogrisel can confirm.

Reply:
> Do I assume right, that we don't so anything about it at the moment?

I think you can ignore these warnings.

For now we turn some warnings into errors like `DeprecationWarning` and `FutureWarning` in the CI and locally when setting `SKLEARN_WARNING_AS_ERRORS`. @jeremiedbb has a master plan to try to pay more attention to warnings in general see https://github.com/scikit-learn/scikit-learn/pull/29516